src/util.js

```diff
@@ -0,0 +1,32 @@
+'use strict'
+
+exports.first = async (iterator) => {
+  for await (const value of iterator) {
+    return value
+  }
+}
+
+exports.last = async (iterator) => {
+  let value
+  for await (value of iterator) {
+    // Intentionally empty
+  }
+  return value
+}
+
+exports.all = async (iterator) => {
+  const values = []
+  for await (const value of iterator) {
+    values.push(value)
+  }
+  return values
+}
+
+exports.fancyIterator = (next) => {
+  const iterator = { next }
+  iterator[Symbol.asyncIterator] = function () { return this }
+  iterator.first = () => exports.first(iterator)
+  iterator.last = () => exports.last(iterator)
+  iterator.all = () => exports.all(iterator)
+  return iterator
+}
```

Comment:
The reason this is a decorator in IPFSX is because [generators have `return` and `throw` functions](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Generator) that are called automatically by a `for await` loop to allow the generator to clean up when execution is aborted/finished.

```js
const iterator = {
  return () {
    console.log('return called')
    return { done: true }
  },
  next () {
    return { done: false, value: 1 }
  },
  [Symbol.iterator] () {
    return iterator
  }
}

async function main () {
  let i = 0
  for await (const value of iterator) {
    i++
    if (i === 3) break
    console.log(value)
  }
}

main()

// output:
// 1
// 1
// return called
```

If we create a custom async iterator here that isn't a generator then breaking out of a loop that uses it will not call return, all that will happen is that `next` will not be called again. I don't think this is a problem _right now_ with the code in this PR...but there could be issues here e.g. `fs.createReadStream` with a transform that converts raw data into an IPLD node, passed to `ipld.put`, halting half way would never end the stream, could cause the whole of the file content to be buffered into memory.

Reply:
With "being a decorator" you mean that you're extending/modifying the iterator you passed in? I previously had that until @achingbrain mentioned that it's bad style to modify things you pass in (I agree with that). But in this case it might make sense.

What if fancyiterator just implements a return and throw? I know it won't be general purpose but good enough for js-ipld, won't it?

Reply:
Typically a [decorator](https://en.wikipedia.org/wiki/Decorator_pattern) would wrap the instance being decorated instead of modifying it, but you could argue it's a side-effect of a type system that doesn't allow modification at runtime.